lib/auth/auth_with_roles.go

```diff
@@ -1483,6 +1483,108 @@ func (a *ServerWithRoles) GetNode(ctx context.Context, namespace, name string) (
 	return node, nil
 }
 
+// ListUnifiedResources returns a paginated list of unified resources filtered by user access.
+func (a *ServerWithRoles) ListUnifiedResources(ctx context.Context, req *proto.ListUnifiedResourcesRequest) (*types.ListResourcesResponse, error) {
+	// Fetch full list of nodes in the backend.
+	startFetch := time.Now()
+	unifiedResources, err := a.authServer.unifiedResourceWatcher.GetUnifiedResources(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	elapsedFetch := time.Since(startFetch)
+
+	// Filter resources to return the ones for the connected identity.
+	filteredResources := make([]types.ResourceWithLabels, 0)
+	startFilter := time.Now()
+	for _, resource := range unifiedResources {
+		switch r := resource.(type) {
```

Comment:
This code is important from the security standpoint. I'd like to have a test case that makes sure that inaccessible resources _of all types_ cannot be fetched.